### DISCUSSION TITLE
Magick++ remap giving different output that CLI convert

### DISCUSSION THREAD

**Discussion by mickes27** [ImageMagick/ImageMagick#6913] [Help]
### ImageMagick version

7.1.1-21 Q16-HDRI x64

### Operating system, version and so on

Windows 10 64-bit

### Description

Hey. I want to reduce colors of an image to predefined 24-colored palette. For testing purposes I am using this image:

Original:
![Tiger](https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEi161RF2vLtTfe3j0q7Yh2jogrO_tLyDZ2a_p2h71NnyzBd_qUHCtlztYVjpZRzDnRTPgar_6OUxN9IWP5aYC_aXor8aQXdUmAUKKl810KKz8WjPeUPhj_EiOHnQtMCyi6qQJaOuwur95VqOoz6QTCZn3zmrVonyTJkPay0EYwlW8KCom0LsSskdJ5H/s1024/wogoco1797_a_roaring_bengal_tiger_in_jungle_hyper_realstic_8k_cafd1714-d946-4060-a2a7-9e2de9c4acad.png)

From CLI I am using:
`convert Tiger.png -dither FloydSteinberg -remap Palette.png Tiger_dither_remap_manual.png`

Which gives output:
![Tiger_dither_remap_manual](https://github.com/ImageMagick/ImageMagick/assets/5618183/926ae757-d4be-4308-ab48-2cbf8f0b3788)

From C++ using Magick++ I have:

```
Magick::Image img("Tiger.png");
Magick::Image remap("Palette.png");
img.quantizeDitherMethod(Magick::DitherMethod::FloydSteinbergDitherMethod);
img.quantize();
img.map(remap, false);
      
img.write("Tiger_dither_remap_cpp.png");
```

Which gives me:
![Tiger_dither_remap_cpp](https://github.com/ImageMagick/ImageMagick/assets/5618183/da023d70-2efc-478e-aac2-21f94184a892)

Also, without `Image::quantize()` methods and using `remap(remap, true)` the result is not the same as from CLI. Am I doing something wrong? What Magick++ API should I use to obtain result the same as from CLI?

EDIT: Changed original image link and added Palette.png:
![Palette](https://i.ibb.co/ThS87k4/Palette.png)

**Comment by snibgo** [comment]
Please show your `Palette.png`.

EDIT: I can't download your tiger image, perhaps because the URL has "private-user".

**Reply by mickes27** [reply]
I updated original image and added Palette.png. Hope it works this time

**Comment by snibgo** [comment]
Thanks, I can download your images.

I'm not sure why you are doing `img.quantize();` in your CPP code. Do you want to quantize, and then remap? Your CLI doesn't do that.

However, there is a known limitation with `FloydSteinbergDitherMethod` in Magick++. See https://github.com/ImageMagick/ImageMagick/discussions/6017

**Reply by mickes27** [reply]
So applying FloydSteinberg dithering first and then calling `map(palette, false)` gives different outcome then calling `map(palette, true)`? Do we have any plans to add FloydSteinberg dithering method for `Image::map`?

**Comment by snibgo** [comment]
> So applying FloydSteinberg dithering first and then calling map(palette, false) gives different outcome then calling map(palette, true)?

Yes, we should expect different results.

In the first case, you are quantizing (reducing the number of colours) which changes pixels. And then you are remapping, which change the pixels again, but based on the quantized colours.

In the second case, you are directly remapping the input image.

> Do we have any plans...

I don't know.
